packages/wrangler/src/__tests__/output.test.ts

```diff
@@ -0,0 +1,229 @@
+import { readdirSync, readFileSync } from "node:fs";
+import { join } from "node:path";
+import { clearOutputFilePath, writeOutput } from "../output";
+import { runInTempDir } from "./helpers/run-in-tmp";
+import type { OutputEntry } from "../output";
+
+const originalProcessEnv = process.env;
+const {
+	WRANGLER_OUTPUT_FILE_DIRECTORY: _,
+	WRANGLER_OUTPUT_FILE_PATH: __,
+	...processEnvNoVars
+} = originalProcessEnv;
+
+describe("writeOutput()", () => {
+	runInTempDir({ homedir: "home" });
+	afterEach(clearOutputFilePath);
+
+	it("should do nothing with no env vars set", () => {
+		try {
+			process.env = processEnvNoVars;
+			writeOutput({
+				type: "wrangler-session",
+				version: 1,
+				wrangler_version: "0.0.0.0",
+				command_line_args: ["--help"],
+			});
+			// No files written
+			expect(readdirSync(".")).toEqual(["home"]);
+		} finally {
+			process.env = originalProcessEnv;
+		}
+	});
+
+	it("should write to the file given by WRANGLER_OUTPUT_FILE_PATH", () => {
+		try {
+			const WRANGLER_OUTPUT_FILE_PATH = "output.json";
+			process.env = { ...processEnvNoVars, WRANGLER_OUTPUT_FILE_PATH };
+			writeOutput({
+				type: "wrangler-session",
+				version: 1,
+				wrangler_version: "0.0.0.0",
+				command_line_args: ["--help"],
+			});
+			const outputFile = readFileSync(WRANGLER_OUTPUT_FILE_PATH, "utf8");
+			expect(outputFile).toContainEntries([
+				{
+					type: "wrangler-session",
+					version: 1,
+					wrangler_version: "0.0.0.0",
+					command_line_args: ["--help"],
+				},
+			]);
+		} finally {
+			process.env = originalProcessEnv;
+		}
+	});
+
+	it("should write to the file given by WRANGLER_OUTPUT_FILE_PATH, ignoring WRANGLER_OUTPUT_FILE_DIRECTORY", () => {
+		try {
+			const WRANGLER_OUTPUT_FILE_PATH = "output.json";
+			process.env = {
+				...processEnvNoVars,
+				WRANGLER_OUTPUT_FILE_PATH,
+				WRANGLER_OUTPUT_FILE_DIRECTORY: ".",
+			};
+			writeOutput({
+				type: "wrangler-session",
+				version: 1,
+				wrangler_version: "0.0.0.0",
+				command_line_args: ["--help"],
+			});
+			const outputFile = readFileSync(WRANGLER_OUTPUT_FILE_PATH, "utf8");
+			expect(outputFile).toContainEntries([
```

Comment:
I think custom matchers are great for common use-cases to a whole codebase and should be introduced globally so they can be defined clearly, learnt and relied upon. Implementing them as one-offs per suite I think is obfuscating

Reply:
It just a helper function to avoid duplicating the same logic throughout and keeping the tests themselves nice and readable.

It is only used in this file (where it is defined) and it makes no sense outside of this file. It is nicer than using a vanilla function because Vitest treats it specially, for example it will format the errors nicely and do diffing of the actual and expected results.

Reply:
If you Cmd-Click on the function it takes you to the type definition below, which is exactly next to the implementation. So I don't think it is too hard to understand.

Reply:
Somewhat ironically, when you create global custom matchers like this it is common to put the type in a `.d.ts` file - e.g. https://github.com/cloudflare/workers-sdk/blob/f6300a71adb83e5f92cd9c9ddd56a85132e03f98/packages/create-cloudflare/vitest.d.ts

In this case it is actually much harder to find the original implementation of the matcher. For example the `toExist()` matcher implementation is over here: https://github.com/cloudflare/workers-sdk/blob/cc335ee1c1dbd58009714c4b49e23e9dcf121d1e/packages/create-cloudflare/e2e-tests/setup.ts